### DISCUSSION TITLE
Retro-Assembly Docker Container

### DISCUSSION THREAD

**Discussion by thewhiterabbitx** [arianrhodsandlot/retroassembly#10] [Ideas]
This project is fantastic, from the ground up! Amazing UI and amazing functionalities.

I'm not a developer, but I like to self-host. This project would be a perfect subject for a containerized solution. Just spin up your container, place your ROMs in a binded host folder and connect your browser to a specified port on the running machine.

It would be the future of Retro-Pie.

(If a docker solution is not on the table, could it be possible to have a step guide for dummies on how to self-host? I would still love to try to get it running on a NginX container.)

**Comment by arianrhodsandlot** [comment]
Unfortunately, this project is not made for self hosting for now. If you self-host an instance, you will only be able to select a local directory as the game source, and cloud services won't work anymore. If you do not care about that, I just created a docker image for test usage, try `docker pull arianrhodsandlot/retro-assembly`.

However, I do think your suggestion is a great idea. I'd like to spend some time on adding support for self-hosting in the future... when I have time. Start a docker server and give the server a ROM directory, then we can play with this self-hosted instance on any devices, that's an ideal use case.

**Reply by thewhiterabbitx** [reply]
Wow! Pulled and running! Will do some testing in the upcoming days... This is a game changer, zero-touch setup, play anywhere on your network, even beyond your network if TailScale or any VPN solution is running.

Yes, definitely the ENV parameter to bind a local directory needs to be part of the MVP, but man, only thinking to be able running it like this it's fabulous!

Not sure how to load local directory though, the "Continue with your own library" does not look like working on my browser.

Thanks!

**Reply by arianrhodsandlot** [reply]
Are you using Firefox or an iPhone
?

Since cloud services like are all unavailable in self hosted instances, the only left way to play custom roms is selecting a local directory. But unfortunately, not all browsers support selecting a directory. So if you're in this case, there won't be any way to import ROMs.

As i know, only Chrome(but not including Chrome for iOS) and Chromium based browsers support selecting a directory.

**Reply by thewhiterabbitx** [reply]
Yeah, browser problem... I'm using Brave on MacOS and it's not working (though I believe Brave is Chromium based - possibly some security feature to disable). Installed Chrome and working fine OOTB!

Already playing! With this docker image it's 10/10 the best retro-gaming set-up experience! Unbelievable how many hours this would spare worldwide. sorry but I'm just super stoked.

Thanks a lot!

**Reply by thewhiterabbitx** [reply]
And yes, passing the rom directory to the container would allow to solve the compatibility problem with any browser...

I tried with few directory/file overwrite hoping to substitute the "public library" ROMs, but the structure is already created with the docker build so the overwrite does not help...

Will wait for the future.

Thanks again!

**Reply by Overview3833** [reply]
> Unfortunately, this project is not made for self hosting for now. If you self-host an instance, you will only be able to select a local directory as the game source, and cloud services won't work anymore. If you do not care about that, I just created a docker image for test usage, try `docker pull arianrhodsandlot/retro-assembly`.
> 
> However, I do think your suggestion is a great idea. I'd like to spend some time on adding support for self-hosting in the future... when I have time. Start a docker server and give the server a ROM directory, then we can play with this self-hosted instance on any devices, that's an ideal use case.

Where can I find instructions for setting parameters？
like:
```
port
rom path
```

**Comment by Overview3833** [comment]
Okey,I found the port 3000 at docker logs
However, I have tested multiple chromium browsers but they cannot upload roms correctly from local. Is there any reference method to store roms directly on my self-hosted server for local reading?

**Reply by arianrhodsandlot** [reply]
Unfortunately self-hosted ROMs are not supported yet. It's still waiting for being implemented.

**Reply by muditjuneja** [reply]
@arianrhodsandlot if you have some notes/plan around the implementation, I can start looking into that! This would be a really nice addition to the stack in general as well.

**Comment by arianrhodsandlot** [comment]
Just try https://hub.docker.com/r/arianrhodsandlot/retroassembly

**Reply by thewhiterabbitx** [reply]
Amazing! Testing right now! Till now flawless!
